File: src/id3v2/tag.rs
use winsafe::{self as w};

use crate::id3v2::*;

/// Metadata of a single MP3 file.
#[derive(Default, Clone)]
pub struct Tag {
	mp3_offset: usize,
	padding: usize,
	frames: Vec<Frame>,
}

impl std::fmt::Display for Tag {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(
			f,
			"Off: {}, pad: {}\n{}",
			self.mp3_offset,
			self.padding,
			self.frames
				.iter()
				.map(|f| f.to_string())
				.collect::<Vec<_>>()
				.join("\n"),
		)
	}
}

impl Tag {
	#[must_use]
	pub const fn padding(&self) -> usize {
		self.padding
	}
	#[must_use]
	pub const fn frames(&self) -> &Vec<Frame> {
		&self.frames
	}
	#[must_use]
	pub const fn frames_mut(&mut self) -> &mut Vec<Frame> {
		&mut self.frames
	}

	/// Reads the tag from an MP3 file.
	#[must_use]
	pub fn read_from_file(mp3_path: &str) -> w::AnyResult<Self> {
		let fin = w::FileMapped::open(mp3_path, w::FileAccess::ExistingReadOnly)?;
		Self::parse(fin.as_slice())
	}

	/// Parses the tag from a binary blob.
	#[must_use]
	pub fn parse(src: &[u8]) -> w::AnyResult<Self> {
		match Self::parse_header(src)? {
			Some(_declared_size) => {
				// MP3 file has an ID3v2 tag.
				let (frames, mp3_offset, padding) = Self::parse_frames(&src[10..])?;
				Ok(Self { mp3_offset, padding, frames })
			},
			None => Ok(Self::default()), // MP3 file has no ID3v2 tag
		}
	}

	/// If an ID3v2 tag is present, return its declared size, including the
	/// 10-byte header.
	#[must_use]
	fn parse_header(src: &[u8]) -> w::AnyResult<Option<u32>> {
		// Check ID3 magic bytes.
		let (magic_bytes, _) = util::parse_ascii(src, 3);
		if magic_bytes != "ID3" {
			return Ok(None); // MP3 file has no tag
		}

		// Validate tag version 2.3.0.
		// The first "2" is not stored in the tag.
		if src[3..5] != [3, 0] {
			return Err(format!(
				"Tag version 2.{}.{} is not supported, only 2.3.0.",
				src[3], src[4],
			)
			.into());
		}

		// Validate unsupported flags.
		if src[5] & 0b1000_0000 != 0 {
			return Err("Unsynchronised tag not supported.".into());
		} else if src[5] & 0b0100_0000 != 0 {
			return Err("Tag extended header not supported.".into());
		}

		// Read declared tag size; also count 10-byte tag header.
		let declared_size = util::synchsafe_decode(u32::from_be_bytes(src[6..10].try_into()?)) + 10;
		Ok(Some(declared_size))
	}

	/// Returns the frames, MP3 offset and padding size.
	#[must_use]
	fn parse_frames(mut src: &[u8]) -> w::AnyResult<(Vec<Frame>, usize, usize)> {
		let mut frames = Vec::with_capacity(10); // arbitrary
		let mut offset = 10usize; // start at 10 because src already skipped 10-byte header

		// Known magic byte sequences that identify the beginning of a MP3.
		// https://stackoverflow.com/a/7302482/6923555
		// https://en.wikipedia.org/wiki/List_of_file_signatures
		// https://github.com/sindresorhus/file-type/issues/75#issuecomment-320650344
		const MP3_MAGIC: [[u8; 2]; 5] =
			[[0xff, 0xfb], [0xff, 0xfb], [0xff, 0xf2], [0xff, 0xfa], [0xff, 0xf3]];

		loop {
			if MP3_MAGIC.iter().any(|magic| magic == &src[0..2]) {
				// We found the beginning of the MP3 file, no padding.
				return Ok((frames, offset, 0));
			} else if src[0] == 0x0000 {
				// We entered a padding region after all frames.
				match src
					.windows(2)
					.position(|by| MP3_MAGIC.iter().any(|magic| magic == by))
				{
					Some(idx_mp3_start) => {
						// We found the beginning of the MP3 file, after some padding.
						return Ok((frames, offset + idx_mp3_start, idx_mp3_start));
					},
					None => return Err("MP3 offset not found.".into()),
				}
			}

			let (new_frame, declared_size) = Frame::parse(src)?;
			if declared_size > src.len() {
				// Means the size was serialized with error.
				return Err(format!(
					"Frame size is greater than available size: {} vs {}.",
					declared_size,
					src.len(),
				)
				.into());
			}

			offset += declared_size;
			src = &src[declared_size..]; // move past the frame
			frames.push(new_frame); // add the frame to our collection
		}
	}

	/// Serializes the tag into bytes.
	#[must_use]
	pub fn serialize(&self) -> Vec<u8> {
		let sz_tag = self
			.frames
			.iter()
			.fold(10, |acc, frame| acc + frame.serialize_size()); // count 10-byte tag header

		let mut blob = Vec::with_capacity(sz_tag);
		util::serialize_ascii(&mut blob, "ID3"); // magic bytes
		blob.extend([0x03, 0x00]); // tag version 2.3.0
		blob.push(0x00); // flags
		blob.extend(util::synchsafe_encode(sz_tag as u32 - 10).to_be_bytes()); // won't count 10-byte header

		self.frames
			.iter()
			.for_each(|frame| frame.serialize(&mut blob));

		blob
	}

	/// Saves the tag to an MP3 file. If there are no frames, the tag will be
	/// entirely removed from the file.
	///
	/// No padding will be written, and the `padding` field will be set to zero.
	pub fn save_to_file(&mut self, mp3_path: &str) -> w::AnyResult<()> {
		let fout = w::File::open(mp3_path, w::FileAccess::ExistingRW)?;
		let current_contents = fout.read_all()?; // read the whole MP3 into a buffer
		let old_tag = Self::parse(&current_contents)?; // so we can extract MP3 offset

		if self.frames.is_empty() {
			fout.erase_and_write(
				&current_contents[old_tag.mp3_offset..], // no tag will be written
			)?;
		} else {
			fout.erase_and_write(
				&self
					.serialize()
					.into_iter()
					.chain(current_contents[old_tag.mp3_offset..].iter().map(|b| *b)) // MP3 data
					.collect::<Vec<_>>(),
			)?;
		}
		self.padding = 0; // because we write no padding
		Ok(())
	}

	#[must_use]
	pub fn frame_by_name4(&self, name4: &str) -> Option<&Frame> {
		self.frames.iter().find(|frame| frame.name4() == name4)
	}

	/// If the given frame does exist, tries to set the editable string on it.
	///
	/// If the frame doesn't exist, creates a new one with the editable string
	/// on it.
	///
	/// An empty string will remove the frame, if any.
	pub fn set_editable_string(&mut self, name4: &str, val: &str) -> w::AnyResult<()> {
		let val = val.trim();
		if val.is_empty() {
			self.frames.retain(|frame| frame.name4() != name4); // remove empty new values
		} else {
			match self.frames.iter_mut().find(|frame| frame.name4() == name4) {
				Some(frame) => frame.set_editable_str(val)?, // frame already exists, set new value
				None => {
					let new_frame = Frame::new_from_editable_str(name4, val)?;
					self.frames.push(new_frame); // frame doesn't exist, push new
				},
			}
		}
		Ok(())
	}

	/// Any ReplayGain frame present?
	#[must_use]
	pub fn has_replay_gain(&self) -> bool {
		self.frames.iter().any(|frame| {
			if frame.name4() == "TXXX"
				&& let Body::UserText(ut) = frame.body()
				&& ut.descr.starts_with("replaygain")
			{
				true
			} else {
				false
			}
		})
	}
}

File: src/id3v2/Tag.cpp
#include <algorithm>
#include <stdexcept>
#include <windlg/lib.h>
#include "Tag.h"
#include "util.h"
using std::optional, std::span, std::vector, std::wstring_view;
using namespace lib;
using namespace id3;

Tag::Tag(wstring_view mp3)
	: path{mp3}
{
	lib::FileMapped f{mp3, lib::FileMapped::Access::ExistingReadOnly};
	_parseBin(f.asSpan());
}

optional<const Frame*> Tag::frameByName4(wstring_view name4) const
{
	for (const Frame& frame : frames) {
		if (lib::str::eqI(frame.name4, name4))
			return &frame;
	}
	return std::nullopt;
}

optional<Frame*> Tag::frameByName4(wstring_view name4)
{
	auto pFrame = const_cast<const Tag*>(this)->frameByName4(name4);
	return pFrame.has_value() ? optional{const_cast<Frame*>(pFrame.value())} : std::nullopt;
}

LPCWSTR Tag::replayGainStatus() const
{
	bool hasTrack = false;
	bool hasAlbum = false;

	for (const Frame& frame : frames) {
		if (hasTrack && hasAlbum) break;

		if (lib::str::eqI(frame.name4, L"TXXX")) {
			if (auto pData = std::get_if<Frame::UserText>(&frame.data); pData) {
				if (lib::str::startsWithI(pData->descr, L"replaygain_track_"))
					hasTrack = true;
				else if (lib::str::startsWithI(pData->descr, L"replaygain_album_"))
					hasAlbum = true;
			}
		}
	}
	
	if (hasTrack && hasAlbum) return L"TA";
	else if (hasTrack) return L"T";
	else if (hasAlbum) return L"A";
	else return L"";
}

void Tag::saveToFile() const
{
	if (path.empty())
		throw std::runtime_error("Tag has no path");

	lib::File fout{path, lib::File::Access::ExistingRW};
	vector<BYTE> currentContents = fout.readAll();
	HeaderInfo headerNfo = _ParseHeader(currentContents);

	fout.setSize(0);
	if (!frames.empty()) {
		vector<BYTE> tagBlob = _serialize();
		fout.write(tagBlob);
	}
	fout.write({currentContents.begin() + headerNfo.mp3Offset, currentContents.end()}); // MP3 data
}

void Tag::_parseBin(span<BYTE> src)
{
	HeaderInfo headerNfo = _ParseHeader(src);
	if (!headerNfo.declaredSize && !headerNfo.mp3Offset)
		return; // MP3 has no ID3v2 tag

	FramesInfo framesNfo = _ParseFrames(src.subspan(10, headerNfo.mp3Offset - 10));
	mp3Offset = headerNfo.mp3Offset;
	padding = framesNfo.padding;
	frames = std::move(framesNfo.frames);
}

Tag::HeaderInfo Tag::_ParseHeader(span<BYTE> src)
{
	HeaderInfo nfo{};

	// Retrieve MP3 offset.
	optional<size_t> maybeMp3Offset = util::positionOf2(src, 0xff, 0xfb); // https://stackoverflow.com/a/7302482/6923555
	if (!maybeMp3Offset.has_value()) [[unlikely]] {
		throw std::runtime_error("No MP3 signature found");
	}
	nfo.mp3Offset = static_cast<UINT>(maybeMp3Offset.value());

	// Check ID3 magic bytes.
	if ( !(src[0] == 'I' && src[1] == 'D' && src[2] == '3') ) {
		return nfo; // MP3 has no ID3v2 tag
	}

	// Validate tag version.
	if ( !(src[3] == 3 && src[4] == 0) ) { // the "2" of "2.3.0" is not serialized
		throw std::runtime_error( str::toAnsi(str::fmt(L"Tag v2.%d.%d not supported, only v2.3.0", src[3], src[4])) );
	}

	// Validate unsupported flags.
	if (src[5] & 0b1000'0000) {
		throw std::runtime_error("Unsynchronised tag not supported");
	} else if (src[5] & 0b0100'0000) {
		throw std::runtime_error("Tag extended header not supported");
	}

	nfo.declaredSize = util::syncSafe::decode(util::uintFromBeBytes(src.subspan(6, 4)));
#ifdef _DEBUG
	if (nfo.declaredSize > nfo.mp3Offset) {
		auto msg = str::fmt(L"--- Declared size: %d > offset: %d\n", nfo.declaredSize, nfo.mp3Offset);
		OutputDebugStringW(msg.c_str());
	}
#endif

	return nfo;
}

Tag::FramesInfo Tag::_ParseFrames(span<BYTE> src)
{
	FramesInfo nfo{};
	for (;;) {
		if (src.empty()) { // end of tag, no padding found
			break;
		} else if (vec::all(src, 0x00)) { // we entered a padding region after all frames
			nfo.padding = static_cast<UINT>(src.size());
			break;
		}

		Frame frame{src};
		if (frame.declaredSize > src.size()) { // means the size was serialized with error
			throw std::runtime_error(
				str::toAnsi(str::fmt(L"Declared frame size greater than available size: %d vs %d",
					frame.declaredSize, src.size())) );
		}

		src = src.subspan(frame.declaredSize);
		nfo.frames.emplace_back(std::move(frame));
	}
	return nfo;
}

optional<size_t> Tag::_apicSize() const
{
	if (optional<const Frame*> frame = frameByName4(L"APIC"); frame.has_value()) {
		auto pic = std::get_if<Frame::Picture>(&frame.value()->data);
		return pic->bin.size();
	}
	return std::nullopt;
}

vector<BYTE> Tag::_serialize() const
{
	size_t apicSize = _apicSize().value_or(0);

	auto buf = vec::newReserved<BYTE>(10 + 10 * frames.size() + apicSize); // arbitrary
	util::serializeChars(L"ID3", buf); // magic bytes
	vec::append(buf, 0x03, 0x00); // tag version
	vec::append(buf, 0x00); // flags

	size_t offsetSz = buf.size();
	buf.insert(buf.end(), 4, 0x00); // data size placeholder

	size_t szFrames = 0; // won't count 10-byte tag header
	for (const Frame& frame : frames)
		szFrames += frame.serialize(buf);

	util::serializeInPlaceUintBe(util::syncSafe::encode(static_cast<UINT>(szFrames)), buf.begin() + offsetSz);
	return buf;
}
